File: autocontext/src/autocontext/production_traces/validate.py
"""Validation helpers for ``ProductionTrace`` documents.

The generated Pydantic model in ``contract.models`` is the single source of
truth for shape (DRY: never redefined here). These helpers are customer-facing
conveniences:

* :func:`validate_production_trace` — raising variant (idiomatic Pydantic).
* :func:`validate_production_trace_dict` — ergonomic non-raising variant that
  returns ``(ok, errors)``; messages carry a dot-path field pointer flattened
  from Pydantic's structured errors.
"""

from __future__ import annotations

from typing import Any

from pydantic import ValidationError

from autocontext.production_traces.contract.models import ProductionTrace


def validate_production_trace(data: dict[str, Any]) -> ProductionTrace:
    """Validate and parse a production-trace document.

    Raises :class:`pydantic.ValidationError` if the input fails schema validation
    (including any branded-id pattern constraints on the contained fields).
    """
    return ProductionTrace.model_validate(data)


def validate_production_trace_dict(data: Any) -> tuple[bool, list[str]]:
    """Non-raising variant. Returns ``(True, [])`` on valid input; otherwise
    ``(False, [<messages>])`` with one entry per Pydantic error.

    Error messages are formatted as ``"<dot.path>: <message>"`` where
    ``<dot.path>`` is derived from the error ``loc`` tuple — e.g.
    ``"messages.0.role"`` for a bad enum value in the first message.
    """
    try:
        ProductionTrace.model_validate(data)
    except ValidationError as exc:
        return False, [_format_error(err) for err in exc.errors()]
    return True, []


def _format_error(err: Any) -> str:
    loc = err.get("loc") or ()
    path = ".".join(str(p) for p in loc) if loc else "<root>"
    msg = err.get("msg", "invalid")
    return f"{path}: {msg}"


__all__ = [
    "validate_production_trace",
    "validate_production_trace_dict",
]

File: ts/src/production-traces/sdk/validate.ts
import type { ProductionTrace } from "../contract/types.js";
import { validateProductionTrace as validateViaAjv } from "../contract/validators.js";

/**
 * Customer-facing validation surface for ``ProductionTrace`` documents.
 *
 * DRY anchor: both entry points delegate to the AJV validator shipped in
 * ``production-traces/contract/validators.ts``. The JSON Schemas are the
 * single source of truth; this module is a thin ergonomics layer.
 *
 * DDD anchor: names mirror Foundation A Layer 6's Python SDK —
 * ``validate_production_trace`` (throws) + ``validate_production_trace_dict``
 * (non-throwing). camelCase only translates the naming convention; semantics
 * match Python exactly.
 */

/**
 * Structured validation failure. Carries a summary message plus the list of
 * per-field errors that AJV reported. Enterprise integrations typically log
 * ``fieldErrors`` directly for operator visibility.
 */
export class ValidationError extends Error {
  readonly fieldErrors: readonly string[];

  constructor(message: string, fieldErrors: readonly string[]) {
    super(message);
    this.name = "ValidationError";
    this.fieldErrors = fieldErrors;
    // Set prototype so `instanceof` works after transpilation targeting ES5+
    // environments. Node 18+ / modern runtimes honor this pattern.
    Object.setPrototypeOf(this, ValidationError.prototype);
  }
}

/**
 * Ergonomic result shape for the non-throwing validator. ``errors`` is always
 * present (empty array on success) so call sites never need a defined-check.
 */
export interface ValidateResult {
  readonly valid: boolean;
  readonly errors: readonly string[];
}

/**
 * Validate and return a ``ProductionTrace`` document. On failure raises
 * :class:`ValidationError` carrying the structured AJV errors in
 * ``fieldErrors``.
 */
export function validateProductionTrace(input: unknown): ProductionTrace {
  const result = validateViaAjv(input);
  if (result.valid) {
    return input as ProductionTrace;
  }
  const errors = result.errors;
  const message = errors.length === 1
    ? `ProductionTrace validation failed: ${errors[0]}`
    : `ProductionTrace validation failed: ${errors.length} errors (first: ${errors[0]})`;
  throw new ValidationError(message, errors);
}

/**
 * Non-raising variant — returns ``{ valid, errors }``. Mirrors Python's
 * ``validate_production_trace_dict`` for customers who prefer to branch on a
 * flag rather than try/catch.
 */
export function validateProductionTraceDict(input: unknown): ValidateResult {
  const result = validateViaAjv(input);
  if (result.valid) return { valid: true, errors: [] };
  return { valid: false, errors: result.errors };
}
